language-plutus-core/generators/Language/PlutusCore/Generators/Test.hs

```diff
@@ -1,18 +1,116 @@
-import           Data.Foldable
-import           Hedgehog                              hiding (Size, Var, eval)
+import qualified Data.Text.IO                                            as Text
+import           Data.Traversable
+import           Hedgehog                                                hiding (Size, Var, eval)
+import           System.FilePath                                         ((</>))
+
+{- Note [Type-eval checking]
+We generate terms along with values they are supposed to evaluate to. Before evaluating a term,
+we type check it. Then we evaluate the term and check whether the expected result matches with
+the actual one. Thus "type-eval checking".
+-}
+
+-- | The type of errors that can occur during type-eval checking.
+data TypeEvalCheckError
+    = TypeEvalCheckErrorIllFormed (Error ())
+    | TypeEvalCheckErrorIllEvaled (Value TyName Name ()) (Value TyName Name ())
+
+-- | Type-eval checking of a term results in a value of this type.
+data TypeEvalCheckResult = TypeEvalCheckResult
+    { _termCheckResultType  :: NormalizedType TyNameWithKind ()
+      -- ^ The type of the term.
+    , _termCheckResultValue :: EvaluationResult
+      -- ^ The result of evaluation of the term.
+    }
+
+instance (PrettyBy config (Error ()), PrettyBy config (Value TyName Name ())) =>
+        PrettyBy config TypeEvalCheckError where
+    prettyBy config (TypeEvalCheckErrorIllFormed err)             =
+        "The term is ill-formed:" <+> prettyBy config err
+    prettyBy config (TypeEvalCheckErrorIllEvaled expected actual) =
+        "The expected value:" <+> prettyBy config expected <> hardline <>
+        "doesn't match with the actual value:" <+> prettyBy config actual
+
+-- | The monad type-eval checking runs in.
+type TypeEvalCheckM = ExceptT TypeEvalCheckError Quote
+
+-- See Note [Type-eval checking].
+-- | Type check and evaluate a term and check that the expected result is equal to the actual one.
+typeEvalCheckBy
+    :: (Term TyName Name () -> EvaluationResult) -- ^ An evaluator.
+    -> TermOf (TypedBuiltinValue Size a)
+    -> TypeEvalCheckM (TermOf TypeEvalCheckResult)
+typeEvalCheckBy eval (TermOf term tbv) = TermOf term <$> do
+    let typecheck = annotateTerm >=> typecheckTerm (TypeCheckCfg 1000 $ TypeConfig True mempty)
+    termTy <- convertErrors TypeEvalCheckErrorIllFormed $ typecheck term
+    resExpected <- liftQuote $ unsafeMakeBuiltin tbv
+    fmap (TypeEvalCheckResult termTy) . for (eval term) $ \resActual ->
+        if resExpected == resActual
+            then return resActual
+            else throwError $ TypeEvalCheckErrorIllEvaled resExpected resActual
+
+-- | Throw a PLC error.
+errorPlc :: PrettyBy PrettyConfigPlc err => err -> b
+errorPlc = error . docString . prettyPlcCondensedErrorBy debugPrettyConfigPlcClassic
+
+-- | Type check and evaluate a term and check that the expected result is equal to the actual one.
+-- Throw an error in case something goes wrong.
+unsafeTypeEvalCheck
```

Comment:
I feel like we should just be returning `Either`s and turning them into exceptions at the border with `Hedgehog` if that's really what it prefers (does it not have a way of dealing with `Either`s?)

Reply:
All terms are generated well-typed and they're supposed to evaluate correctly as well, so it's really an internal error if something goes wrong and I feel free to throw. I tried at first to be more smart here and use `MonadFail` (because I was calling this from `IO` and `PropertyT m` (which doesn't implement `MonadFail`)), but then I realized I need to call it from `GenT m` which has

```haskell
 MonadThrow m => MonadThrow (GenT m)
```

I.e. you can throw, but not natively and the support for non-`Identity` things in `hedgehog` is rather limited. Too much hassle for something that is never supposed to happen.